### DISCUSSION TITLE
Symmetry Problems in Jet Flow Simulation

### DISCUSSION THREAD

**Discussion by tianninglyu** [PrincetonUniversity/athena#514] [Q&A]
Hello everyone,

I hope this message finds you well. I have a specific concern regarding the GetReal function in Athena++ when working with negative numbers, particularly in the context of simulating jets.

Currently, I am conducting simulations of jets using Athena++ and have encountered an unexpected behavior. When I set the jet's magnetic field to both 0.1 and -0.1, I obtained identical simulation results. This outcome has raised doubts about the GetReal function's ability to correctly handle negative values.

Thank you very much for your attention and support. I am grateful for any assistance or suggestions you can provide regarding this issue.

Additionally, my timezone just entered June 1st. I would like to wish all athena++ developers and users a happy Children's Day, if you want your childlike innocence to last forever.

The focus of the question has changed. Please read below.

**Comment by felker** [comment]
`GetReal` can certainly handle negative values just fine--- which block and parameter are you setting in the input file?

**Reply by tianninglyu** [reply]
Here is my input file:
```
<comment>
problem   = jet
reference = 
configure = --prob=jet

<job>
problem_id = jet      # problem ID: basename of output filenames


<output1>
file_type  = hst      # History data dump
dt         = 0.1      # time increment between outputs

<output2>
file_type  = hdf5      # Binary data dump
variable   = prim     # variables to be output
dt         = 1.0      # time increment between outputs

<time>
cfl_number = 0.4      # The Courant, Friedrichs, & Lewy (CFL) Number
nlim       = -1       # cycle limit
tlim       = 300.0    # time limit
integrator  = vl2       # time integration algorithm
xorder      = 2         # order of spatial reconstruction
ncycle_out  = 1         # interval for stdout summary info

<mesh>
nx1        = 300      # Number of zones in X1-direction
x1min      = 0.0      # minimum value of X1
x1max      = 60.0     # maximum value of X1
ix1_bc     = user     # inner-X1 boundary flag
ox1_bc     = outflow  # outer-X1 boundary flag

nx2        = 250      # Number of zones in X2-direction
x2min      = -25.0    # minimum value of X2
x2max      = 25.0     # maximum value of X2
ix2_bc     = outflow  # inner-X2 boundary flag
ox2_bc     = outflow  # outer-X2 boundary flag

nx3        = 1        # Number of zones in X3-direction
x3min      = -25.0    # minimum value of X3
x3max      = 25.0     # maximum value of X3
ix3_bc     = outflow  # inner-X3 boundary flag
ox3_bc     = outflow  # outer-X3 boundary flag

<meshblock>
nx1 = 50
nx2 = 50

<hydro>
gamma   = 1.666666666667     # gamma = C_p/C_v

<problem>
d  = 1.0
p  = 6.0e-3
vx = 0.0
vy = 0.0
vz = 0.0
bx = 0.0
by = 0.0
bz = 0.0

djet = 0.1
pjet = 6.0e-3
vxjet = 1.0
vyjet = 0.0
vzjet = 0.0 
bxjet = 0.1
byjet = 0.0
bzjet = 0.0

rjet = 1.0
```

**Reply by tianninglyu** [reply]
emm I'm sorry the code block doesn't seem to be displaying properly...

**Reply by felker** [reply]
I fixed the formatting; you need to use triple backticks on their own lines to fence the code block

**Reply by tianninglyu** [reply]
Very grateful. Also another question, I noticed that athinput.jet is in the hydro path, does that mean you didn't test with the magnetic field enabled?

**Reply by tianninglyu** [reply]
In fact, I ran the tests with the magnetic field set to 0.1 and -0.1, respectively, to verify whether the asymmetry of the simulation results was due to the accumulation of errors, or whether it should be physically asymmetric.

**Reply by felker** [reply]
I am not super familiar with the `pgen/jet.cpp`, but I believe the input file is in `hydro/` because the magnetic field settings it exposes are very limited--- you can set the ambient B field by `bx,by,bz` or the **inlet ghost zone x1 boundary** B field by `bx_jet,by_jet,bz_jet`

**Reply by tianninglyu** [reply]
Got it, that's just I'm doing. When configuring, I use HLLE riemann solver ```--flux=hlle```, and enable magnetic field ```-b```.

**Reply by felker** [reply]
I definitely notice different results depending on the sign of `bxjet` even just looking at the `.hst` file when using HLLD

**Reply by felker** [reply]
And HLLE. Running `time/tlim=50`, it takes 1285 cycles if bxjet=0.1 and 1341 cycles if it is negative

**Reply by tianninglyu** [reply]
OK, I'll try it with hlld right away. Then I use hlle to check again to see if I was careless just now. Thanks!

**Reply by tianninglyu** [reply]
Sorry I can't connect to the supercomputer properly right now so I can't check it immediately. I'll report back as soon as I've checked it out. Thanks again!

**Reply by tianninglyu** [reply]
OK I believe you're right. I checked the running log and found that the number of cycles is indeed different. But this is a more puzzling problem. The impact of changing B from 0.1 to -0.1 is not even as significant as changing B from 10e-10 to 10e-20, which led me to mistakenly think that they are exactly the same after watching the yt visualization image... so weird. Anyway sorry for wasting your time! Thanks!

**Reply by tianninglyu** [reply]
![屏幕截图 2023-06-01 021418](https://github.com/PrincetonUniversity/athena/assets/130236675/f7b55960-c4a4-4494-be3b-554dd9d4bc98)
Here is the image. Both are frame 200. Bx is on the image, By, Bz are all zero.

**Reply by felker** [reply]
Well you are changing the plasma beta of the inflow boundary condition by 20 orders of magnitude in the bottom row, and not changing it at all in the top row. The direction of the B field at the boundary condition does not matter as much as the magnitude in this case, I think, since the ambient field is zero.

**Reply by tianninglyu** [reply]
Ohhh! Thanks a lot! The ambient field is indeed zero. So I'm still confused about how the asymmetry arises. This appears to be a symmetrical situation physically. In one simulation, I increased the resolution to 1200*1000. Instead of solving the asymmetry problem, it caused KH instability, as shown in the figure below.
![pseudocolor_00200](https://github.com/PrincetonUniversity/athena/assets/130236675/480b9bb9-db8d-4ff0-ac6e-f5fac8b29e7f)
Frame 200, Bx=0.1, hlle riemann solver

**Reply by felker** [reply]
Not sure. I would review this: https://github.com/PrincetonUniversity/athena/wiki/Coordinate-Systems-and-Meshes#symmetry-preservation-in-athena

**Reply by tianninglyu** [reply]
I have not noticed this part in the wiki yet. I would review this too. It seems like I should set the JetInner BC on the x2 boundary and let jet transfer along the y-axis? Because it requires plane of symmetry located at ```x1=0.0```. But in my case, plane of symmetry, I think, located at ```x2=0.0```.

**Reply by tianninglyu** [reply]
Another problem. Is this simulating a impulsive jet or a continuous jet? Because by reading jet.cpp, I seem to only understand the setting of the boundary conditions, but I still seem to know nothing about how the code makes the jet inflow.

**Reply by felker** [reply]
If the boundary condition function does not have some explicit time dependence, then it is enforced at all timesteps. So it is a continuous jet forced by constant density, pressure, and vx in the invisible ghost cells to the left of the domain (within the radius of the inflowing jet):
```
djet = 0.1
pjet = 6.0e-3
vxjet = 1.0
```

Is symmetry broken if you turn off magnetic fields?

I can't recall if `x1` is special or not; you could try moving the boundary condition and domain size as you suggested.

**Reply by tianninglyu** [reply]
If turning off magnetic field, or turning on it but set all zero in the input file, the symmetry is very good, not broken. However I think there is maybe a mistake here. If I set the magnetic field as:
```
B_amb = (0,0,0)
B_jet = (0.1,0,0)
```
Are such magnetic fields illegal? because its normal component at the x1 boundary is discontinuous.

**Reply by c-white** [reply]
That boundary condition is valid, it just means there's a magnetic monopole in the boundary. This may mean the ghost cells are not physically realizable, but that's not uncommon with boundary conditions in MHD problems. The monopole can't advect onto the grid proper. You could always try to write a monopole-free boundary condition, but it's tricky (it will have to depend on the time-evolving state of the active cells), and in the end it wouldn't meaningfully change the results.

**Reply by tianninglyu** [reply]
Thank you for your reply!

**Comment by RenyueCen** [comment]
Hi Chris (and Kyle), thanks for helping. Tianning is learning Athena++ to do some experiments on how radio jets propagate through environments around SMBH to large scales. But he is running into the problem of jet becoming asymmetric with time for an apparent symmetric problem using an existing setup routine (in the hydro dir, though). Not sure what is going on. Any help is highly appreciated.

**Reply by tianninglyu** [reply]
Thank you for your kindness and consideration!
